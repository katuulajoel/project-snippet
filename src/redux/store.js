import { applyMiddleware, createStore } from "redux";
import reduxThunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";

import monitorReducersEnhancer from "../configs/enhancers/monitorReducer";
import loggerMiddleware from "../configs/middleware/logger";
import rootReducer from "./reducers";

export function configureStore(preloadedState = {}) {
  const middlewares = [loggerMiddleware, reduxThunk];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const enhancers = [middlewareEnhancer, monitorReducersEnhancer];
  const composedEnhancers = composeWithDevTools(...enhancers);

  const store = createStore(rootReducer, preloadedState, composedEnhancers);

  if (process.env.NODE_ENV !== "production" && module.hot) {
    module.hot.accept("./reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
}

const store = configureStore();
export default store;
