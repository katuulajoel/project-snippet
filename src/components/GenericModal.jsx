/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React, { useRef, useState, useEffect } from "react";
import { Provider } from "react-redux";
import { LOCATION_CHANGE } from "react-router-redux";
import { confirmable } from "react-confirm";
import { Modal, ModalBody, ModalFooter } from "reactstrap";
import styled from "styled-components";
import { useDispatch } from "react-redux";

/* -------------------------- Internel Dependencies ------------------------- */
import Button from "./Button";
import ModalHeader from "./ModalHeader";
import store from "../store";

/* const GenericModal = (props) => {
  const dispatch = useDispatch();
  const [response, setResponse] = useState(null);
  const wrapperRef = useRef(null);

  const { show, proceed, dismiss, cancel, options, modalContent, modalHeader } =
    props;
  let safe_options = options || {};

  useEffect(() => {
    dispatch({ type: LOCATION_CHANGE });
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  const onResponseChange = (e) => {
    setResponse(e.target.value);
  };

  const renderModalContent = () => {
    return typeof modalContent === "string"
      ? modalContent
      : React.cloneElement(modalContent, { proceed, dismiss, cancel });
  };

  const renderModalHeader = () => {
    return modalHeader ? (
      React.cloneElement(modalHeader, { proceed, dismiss, cancel })
    ) : options.title ? (
      <ModalHeader {...props} />
    ) : null;
  };

  const handleClickOutside = (event) => {
    let safe_options = options || {};
    if (wrapperRef && !wrapperRef.current.contains(event.target)) {
      safe_options.dismissAsCancel ? cancel() : dismiss();
    }
  };

  return (
    <Provider store={store}>
      <Modal
        style={safe_options.style}
        isOpen={show}
        toggle={dismiss}
        className={`${safe_options.size ? `modal-${safe_options.size}` : ""} ${
          safe_options.className || ""
        }`}
        backdrop={
          safe_options.hideBackdrop
            ? false
            : safe_options.mustRespond
            ? "static"
            : true
        }
        keyboard={!safe_options.mustRespond}
      >
        <div ref={wrapperRef}>
          {(!options.mustRespond || safe_options.title) && renderModalHeader()}
          <ModalBody>
            <div>{renderModalContent()}</div>
            {safe_options.isPrompt && (
              <div className="form-group">
                <textarea onChange={(e) => onResponseChange(e)} />
              </div>
            )}
          </ModalBody>
          {!safe_options.hideActions && (
            <StyledModalFooter>
              {!safe_options.hideCancel && (
                <Button
                  className="cancel"
                  onClick={() => cancel()}
                  variant="secondary"
                >
                  {safe_options.cancel || "Cancel"}
                </Button>
              )}
              <Button
                className="okay"
                {...(safe_options.form || "")}
                onClick={() => {
                  if (safe_options.isPrompt) {
                    if (response) {
                      proceed(response);
                    }
                  } else {
                    if (!safe_options.form) {
                      proceed();
                    }
                  }
                }}
                disabled={safe_options.isPrompt && !response}
              >
                {safe_options.ok || "OK"}
              </Button>
            </StyledModalFooter>
          )}
        </div>
      </Modal>
    </Provider>
  );
};

const StyledModalFooter = styled(ModalFooter)`
  padding-bottom: 20px;
  padding-top: 20px;
  background: rgba(237, 241, 247, 0.25);

  button {
    box-sizing: border-box;
    border-radius: 4px;
    box-shadow: none;
    font-weight: 600;
    font-size: 16px;
    line-height: 19px;
    margin: 0;

    &.cancel {
      border: 1px solid #3e4857;
      color: #3e4857;
      background-color: #fff;
      margin-right: 16px;
    }

    &.okay {
      color: #fff;
      background-color: #da3451;
      border-color: #da3451;
    }
  }
`;

GenericModal.propTypes = {
  show: PropTypes.bool,
  proceed: PropTypes.func,
  cancel: PropTypes.func,
  dismiss: PropTypes.func,
  modalContent: PropTypes.any,
  options: PropTypes.object,
  modalHeader: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.elementType,
    PropTypes.object,
  ]),
};

GenericModal.defaultProps = {
  modalHeader: null,
};

export default confirmable(GenericModal); */
