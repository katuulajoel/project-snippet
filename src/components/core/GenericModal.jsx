/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React from "react";
import { Provider } from "react-redux";
import { LOCATION_CHANGE } from "react-router-redux";
import { confirmable } from "react-confirm";
import { Modal, ModalBody, ModalFooter } from "reactstrap";
import styled from "styled-components";

/* -------------------------- Internel Dependencies ------------------------- */
import Button from "./Button";
import TextArea from "./TextArea";
import store from "../../store";
// import ModalHeader from "../../dashboard/projects/common/modals/Header";

class GenericModal extends React.Component {
  static propTypes = {
    show: PropTypes.bool,
    proceed: PropTypes.func,
    cancel: PropTypes.func,
    dismiss: PropTypes.func,
    modalContent: PropTypes.any,
    options: PropTypes.object,
    modalHeader: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.elementType,
      PropTypes.object,
    ]),
  };

  static defaultProps = {
    modalHeader: null,
  };

  constructor(props) {
    super(props);
    this.state = { response: null };
    this.wrapperRef = React.createRef();
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    store.dispatch({ type: LOCATION_CHANGE });
    document.addEventListener("mousedown", this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener("mousedown", this.handleClickOutside);
  }

  toggle() {
    this.setState({
      modalOpen: !this.state.modalOpen,
    });
  }

  onResponseChange(e) {
    this.setState({ response: e.target.value });
  }

  renderModalContent() {
    const { proceed, dismiss, cancel, modalContent } = this.props;

    return typeof modalContent === "string"
      ? modalContent
      : React.cloneElement(modalContent, { proceed, dismiss, cancel });
  }

  //   renderModalHeader() {
  //     const { proceed, dismiss, cancel, modalHeader } = this.props;

  //     return modalHeader ? (
  //       React.cloneElement(modalHeader, { proceed, dismiss, cancel })
  //     ) : this.props.options.title ? (
  //       <ModalHeader {...this.props} />
  //     ) : null;
  //   }

  /**
   * Alert if clicked on outside of element
   */
  handleClickOutside(event) {
    const { dismiss, cancel, options } = this.props;
    let safe_options = options || {};
    if (this.wrapperRef && !this.wrapperRef.current.contains(event.target)) {
      safe_options.dismissAsCancel ? cancel() : dismiss();
    }
  }

  render() {
    const { show, proceed, dismiss, cancel, options } = this.props;
    let safe_options = options || {};

    return (
      <Provider store={store}>
        <Modal
          style={safe_options.style}
          isOpen={show}
          toggle={dismiss}
          className={`${
            safe_options.size ? `modal-${safe_options.size}` : ""
          } ${safe_options.className || ""}`}
          backdrop={
            safe_options.hideBackdrop
              ? false
              : safe_options.mustRespond
              ? "static"
              : true
          }
          keyboard={!safe_options.mustRespond}
        >
          <div ref={this.wrapperRef}>
            {!options.mustRespond || safe_options.title
              ? this.renderModalHeader()
              : null}
            <ModalBody>
              <div>{this.renderModalContent()}</div>
              {safe_options.isPrompt ? (
                <div className="form-group">
                  <TextArea onChange={this.onResponseChange.bind(this)} />
                </div>
              ) : null}
            </ModalBody>
            {safe_options.hideActions ? null : (
              <StyledModalFooter>
                {safe_options.hideCancel ? null : (
                  <Button
                    className="cancel"
                    onClick={() => cancel()}
                    variant="secondary"
                  >
                    {safe_options.cancel || "Cancel"}
                  </Button>
                )}
                <Button
                  className="okay"
                  {...(safe_options.form || "")}
                  onClick={() => {
                    if (safe_options.isPrompt) {
                      if (this.state.response) {
                        proceed(this.state.response);
                      }
                    } else {
                      if (!safe_options.form) {
                        proceed();
                      }
                    }
                  }}
                  disabled={safe_options.isPrompt && !this.state.response}
                >
                  {safe_options.ok || "OK"}
                </Button>
              </StyledModalFooter>
            )}
          </div>
        </Modal>
      </Provider>
    );
  }
}

const StyledModalFooter = styled(ModalFooter)`
  padding-bottom: 20px;
  padding-top: 20px;
  background: rgba(237, 241, 247, 0.25);

  button {
    box-sizing: border-box;
    border-radius: 4px;
    box-shadow: none;
    font-weight: 600;
    font-size: 16px;
    line-height: 19px;
    margin: 0;

    &.cancel {
      border: 1px solid #3e4857;
      color: #3e4857;
      background-color: #fff;
      margin-right: 16px;
    }

    &.okay {
      color: #fff;
      background-color: #da3451;
      border-color: #da3451;
    }
  }
`;

export default confirmable(GenericModal);
