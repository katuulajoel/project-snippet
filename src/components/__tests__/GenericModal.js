import React from "react";
import { Provider } from "react-redux";
import { cleanup, render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import GenericModal from "../GenericModal";
import store from "../../redux/store";

afterEach(cleanup);

describe("GenericModal layout test", () => {
  it("GenericModal component snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <GenericModal />
        </Router>
      </Provider>
    );

    expect(asFragment(<GenericModal />)).toMatchSnapshot();
  });
});
