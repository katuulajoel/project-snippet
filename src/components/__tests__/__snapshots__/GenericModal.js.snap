// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericModal layout test GenericModal component snapshot 1`] = `<DocumentFragment />`;
